design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,340 @@
+            kind: ManuelScaler
+            spec:
+              replica: 3
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: Rollout
+            spec:
+              source:
+                revision: frontend-c8bb659c5 
+```
+
+Once the `frontend` component has been changed, ApplicationConfiguration will trigger
+an upgrade for the application.
+
+1. Create a new workload with the same name of the new component revision.
+2. Check all traits' traitDefinition.
+   - Update all traits who have no revision aware to point to the new workload.
+   - If there are traits who have revision aware and specify to old revision, make sure the workload
+   corresponding to the old revision will still work. 
+3. Periodically do garbage collection for all workloads which don't have any trait specified.
+4. Traits who have revision aware should take care of the garbage collection things for the old revision.
+
+
+### Using revisionName field can specify a fixed component version
+
+When a `ControllerRevision` is specified in ApplicationConfiguration, a component change won't affect
+the running application, only ControllerRevision is created.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - revisionName: frontend-a75588698
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: ManuelScaler
+            spec:
+              replica: 3
+```
+
+In this case, all traits will point to the specified revision of workload. 
+
+Old revisions can also work if any revision aware traits pointing to them.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: traffics.core.oam.dev
+spec:
+  revision:
+    aware: true
+    revisonPath:
+      - "spec.route[*].revision"
+  definitionRef:
+    name: traffics.core.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - revisionName: frontend-a75588698
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: Traffic
+            spec:
+              route:
+                - revision: frontend-a1234 # old revision
+                  percent: 60%
+                - revision: frontend-a75588698 # the new one
+                  percent: 40%
+```
+
+We use `[*]` to represent every array in the field.
+
+
+## Impact to the existing system
+
+1. Add `revisionName` field into AppConfig and this field is mutually exclusive with `componentName`,
+   also `componentName` will be optional if `revisionName` is used.
+2. Add `revision` into TraitDefinition and it will affect workload garbage collection on component upgrade.
+3. Workload instance name will always follow ControllerRevision.
+4. Parameter and ParameterValue need to be treated as a special kind of trait. Values from parameters won't
```

Comment:
I think this proposal is already self-contained. It doesn't conflict with the parameter system which seems to be pretty popular in the OAM community. Namely, a  new `ControllerRevision` is created for a `Component` when its parameters are changed.  A change in `parameterValue` does not affect a `Component` revision. We will address the "uniqueness" of a workload in a different proposal that focuses on the `ApplicationConfiguration` upgrade/verisioning instead of a `Component`.

Reply:
Yeah